### ISSUE TITLE
ValueError when using Generate

### ISSUE THREAD

**Issue by aberja** [gryf/slack-backup#8]
The generate command crashed mid-way thru creating log files with a ValueError.  Looking at the original Slack thread and comparing it to the outputted log file, it appears to have crashed when it tried to process a link to a Google docs spreadsheet stored on a user's Google Drive.  Here is the error message:

2018-05-03 15:59:36,509 INFO: Found configuration file: ./slack-backup.ini
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\Scripts\slack-backup", line 10, in <module>
    command.main()
  File "C:\ProgramData\Anaconda3\lib\site-packages\slack_backup\command.py", line 141, in main
    args.func(args)
  File "C:\ProgramData\Anaconda3\lib\site-packages\slack_backup\command.py", line 56, in generate_raport
    slack.generate_history()
  File "C:\ProgramData\Anaconda3\lib\site-packages\slack_backup\client.py", line 144, in generate_history
    self.reporter.generate()
  File "C:\ProgramData\Anaconda3\lib\site-packages\slack_backup\reporters.py", line 197, in generate
    self.write_msg(messages, log_path)
  File "C:\ProgramData\Anaconda3\lib\site-packages\slack_backup\reporters.py", line 79, in write_msg
    data = self._process_message(message)
  File "C:\ProgramData\Anaconda3\lib\site-packages\slack_backup\reporters.py", line 214, in _process_message
    data = super(TextReporter, self)._process_message(msg)
  File "C:\ProgramData\Anaconda3\lib\site-packages\slack_backup\reporters.py", line 108, in _process_message
    data = processor(msg)
  File "C:\ProgramData\Anaconda3\lib\site-packages\slack_backup\reporters.py", line 233, in _msg_file
    ') ' + msg.file.title, msg.text),
  File "C:\ProgramData\Anaconda3\lib\pathlib.py", line 719, in as_uri
    raise ValueError("relative path can't be expressed as a file URI")
ValueError: relative path can't be expressed as a file URI

**Comment by gryf** [comment]
Hm. It doesn't make sense, since traceback says, that method in `reporters.py` which failed is [_msg_file()](https://github.com/gryf/slack-backup/blob/a57d5085b1fa4bd04e49ccae2a3eb6bdf095fe2b/slack_backup/reporters.py#L224-L234). So there shouldn't be a a link, but filename which should be stored in assets directory by default. If such file is missing for some reason, then `fpath` is set to "does_not_exists" and this string could break `PurePath.as_uri` method. I've fixed that in 007fe04c08c604320a2654c1e378ec6617711416 commit.
If, however, for some reason, link to Google doc was detected as a file, than there is a totally different bug.

**Comment by aberja** [comment]
Your changes were spot on and you are correct the Google Docs file is longer available.  According the line in the generated log file now states: "Joe shared a file: (does_not_exists)".  Generate now runs without any issues. ty

**Comment by gryf** [comment]
Ok, got it. The problem here is, that Slack devs mixed up message of type file where user upload a file which is stored on slack servers with something which basically is a link (or attachment in their nomenclature) both of those kinds are marked as subtype `file_share`.

Silent changes to the slack API are really annoying, since there is no versioning of the api, and it is impossible to track them down without an access to json schema, other than constantly fetching several slack teams with all variety of message types…

Other way could be changing approach from trying to keep up with the running json schema and follow [mcoster/slack-auto-export](https://github.com/mcoster/slack-auto-export) approach and keep original data as a json. This might be tempting to give it a try, but I see couple of issues from top of my head with that: first, slack-auto-export will do only a "snapshot" of the available slack history - there is no merging of previously fetched data. Second, uploaded files (not linked) in messages might be inaccessible after shifting history window in time (unless we download them in time). Chasing API changes is unavoidable, and will result in nasty if-else in the code :C 

I'd like to have more information about this particular message during fetch phase. You can add debug option and redirect the stderr-output to a file, and then search for "Joe shared a file" to find this particular message, like:
```
slack-backup -vv fetch 2> log.txt
```
On Windows you probably have to do:
```
slack-backup -vv fetch >log.txt 2>&1
```

**Comment by aberja** [comment]
I understand the issues with invisible API Changes.  

With respect to keeping the data as json files, I first started using https://github.com/mcoster/slack-auto-export but moved here because it was essential to me that new data was merged with previously fetched data and that the downloaded files linked without any issues.  Personally I like your method of using a database to keep the data rather than keeping the data as json files. 

Here is the relevant log file (with private info replaced with xxx)

Interestingly, it appears that the program found the google doc file, but it does not appear to have downloaded the file as there is no F7ARMB4JU directory with the file.

2018-05-06 10:53:25,635 DEBUG: Message data: {"type": "message", "subtype": "file_share", "text": "<@xxx> shared a file: <https://xxx.slack.com/files/xxx/F7ARMB4JU/17|PT Card Count 9/30/17>", "file": {"id": "F7ARMB4JU", "created": 1506819447, "timestamp": 1506819447, "name": "17", "title": "PT Card Count 9/30/17", "mimetype": "application/vnd.google-apps.spreadsheet", "filetype": "gsheet", "pretty_type": "GDocs Spreadsheet", "user": "xxx", "editable": false, "size": 37583, "mode": "external", "is_external": true, "external_type": "gdrive", "is_public": true, "public_url_shared": false, "display_as_bot": false, "username": "", "url_private": "https://docs.google.com/spreadsheets/d/xxx/edit?usp=drivesdk", "thumb_64": "https://files.slack.com/files-tmb/xxx-F7ARMB4JU-xxx/17_64.png", "thumb_80": "https://files.slack.com/files-tmb/xxx-F7ARMB4JU-xxx/17_80.png", "thumb_360": "https://files.slack.com/files-tmb/xxx-F7ARMB4JU-xxx/17_360.png", "thumb_360_w": 254, "thumb_360_h": 360, "thumb_480": "https://files.slack.com/files-tmb/xxx-F7ARMB4JU-xxx/17_480.png", "thumb_480_w": 339, "thumb_480_h": 480, "thumb_160": "https://files.slack.com/files-tmb/xxx-F7ARMB4JU-xxx/17_160.png", "thumb_720": "https://files.slack.com/files-tmb/xxx-F7ARMB4JU-xxx/17_720.png", "thumb_720_w": 509, "thumb_720_h": 720, "thumb_800": "https://files.slack.com/files-tmb/xxx-F7ARMB4JU-xxx/17_800.png", "thumb_800_w": 800, "thumb_800_h": 1132, "thumb_960": "https://files.slack.com/files-tmb/xxx-F7ARMB4JU-xxx/17_960.png", "thumb_960_w": 678, "thumb_960_h": 960, "thumb_1024": "https://files.slack.com/files-tmb/xxx-F7ARMB4JU-xxx/17_1024.png", "thumb_1024_w": 724, "thumb_1024_h": 1024, "image_exif_rotation": 1, "original_w": 1024, "original_h": 1449, "permalink": "https://xxx.slack.com/files/xxx/F7ARMB4JU/17", "channels": ["xxx"], "groups": [], "ims": [], "comments_count": 0, "has_rich_preview": true}, "user": "xxx", "upload": false, "display_as_bot": false, "username": "xxx", "bot_id": null, "edited": {"user": "xxx", "ts": "1506889631.000000"}, "ts": "1506871034.000043"}
2018-05-06 10:53:25,638 DEBUG: Found external file `https://docs.google.com/spreadsheets/d/xxx/edit?usp=drivesdk'
2018-05-06 10:53:25,638 DEBUG: Message data: {"type": "message", "user": "xxx", "text": "<!everyone> Please provide xxx", "ts": "1506871001.000070"}

And for anyone else following along:

1. it appears that the verbose command in Windows has to come after fetch: 
   
   python "C:\ProgramData\Anaconda3\Scripts\slack-backup" fetch -vv > log.txt 2>&1
   
2. Slack uses user id #, so you need to search for "<@ id #> shared a file"

**Comment by gryf** [comment]
https://imgur.com/a/ztM1GFU

**Comment by gryf** [comment]
Also, we have to add new fields to check - `public_url_shared` and probably `mode` and/or `is_external` - which seems redundant (while the first can take `internal`|`external` and the other is boolean). I guess, there are several other things added like that which makes the consumer for this API really confused.

**Comment by aberja** [comment]
Since this is a public forum, I've removed all private information from the log and replaced them with xxx.  So wherever you see xxx, data has been deleted.  This includes the url to the Google docs file.  However, I can assure you that the file is active and it appears that the script also found it...2018-05-06 10:53:25,638 DEBUG: Found external file `https://docs.google.com/spreadsheets/d/xxx/edit?usp=drivesdk'  But it seems that it never downloaded the file (since no folder name F7ARMB4JU exists)

**Comment by gryf** [comment]
What I meant by saying i'm not logged to any google account was, that if google drive require user to be logged in before accessing the file. Besides. What I've end up with google drive link which passed to requests (or curl, wget whatever) produces only an html content, which probably is useless on client side. Probably - since there will be no direct link to specific content (like regular files), but typically link to edit content on google docs side. As for binary files stored on google drive I can;t figure out how to get direct link to the binary file - the only option is to "share" file with others - so this is pretty same as a link to document on google docs…

Long story short - by using simple tools like curl, wget or using library requests you can't simply download file from google, contrary to simple links from slack servers (even if those require authentication/authorization, we already have those stored in session).

**Comment by gryf** [comment]
This is exactly what I'd like to avoid. Providing a way for certain service will end up with a demand for doing so for the others (dropbox, icloud, sugarsync, you name it). Still, those things are not files in a sense of uploaded by a slack users, but links. I'm curious about the decision to mark them as so by slack devs…

What I can do, is to detect shares which are not files uploaded directly by users, look for the `url_private` key in returned json and make it as an attachment instead… I know that it's far from perfect, but probably is a the best what we can do.

**Comment by aberja** [comment]
I know it's a bit of a manual process, but what about creating the appropriate folder for the file that cannot be downloaded and a separate log file that provides a note to the user with:
  1. URLs of the files that were not downloaded; and 
  2. the folders where the file belongs in?

This way, if the user so chooses, they can manually download the file and put it in the appropriate folder.

**Comment by gryf** [comment]
Sorry for keeping you waiting. With https://github.com/gryf/slack-backup/commit/db8527e9afa76ca264237a415fcb147bb6f0f804 slack-backup gained ability for either providing a file with manual downloadables, or converting them to slack attachment. Please, @aberja let me know, if this fits your needs.

**Comment by aberja** [comment]
Thanks for working on this issue.  My turn to apologize for taking so long to respond.  

I have a question about the raw-dir command.  I inserted under [fetch] `raw-dir = "D:/Documents/Slack/slack-backup/logs/raw-dir/"` but nothing appeared in this directory after I ran fetch command.  Is there something I need to do differently?

Also, when I change the 1st line under [fetch] to `url_file_to_attachment = true` and run the following generate command `python "C:/Program Files (x86)/Microsoft Visual Studio/Shared/Anaconda3_64/Scripts/slack-backup" generate -v -d "D:/Documents/Slack/slack-backup/logs/slacklogs.sqlite" --format text -o Output` I receive the following error message: 

2018-09-25 02:15:14,382 INFO: Found configuration file: ./slack-backup.ini
Traceback (most recent call last):
  File "C:/Program Files (x86)/Microsoft Visual Studio/Shared/Anaconda3_64/Scripts/slack-backup", line 10, in <module>
    command.main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\slack_backup\command.py", line 143, in main
    args.func(args)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\slack_backup\command.py", line 55, in generate_raport
    slack = client.Client(args)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\slack_backup\client.py", line 60, in __init__
    self._url_file_to_attachment = args.url_file_to_attachment
AttributeError: 'Namespace' object has no attribute 'url_file_to_attachment'

I've tested this on two different Slack downloads / generates.

If there is anything further you require, please let me know.

**Comment by gryf** [comment]
Hm. I'll take a look on this later this evening. It might be connected with your environment, but I'm not certain about that.

**Comment by gryf** [comment]
So yeah.
As for raw option. I've failed with providing proper documentation on that option. I'll add a clear docs for that, apologies. In config file it will be an option with underscore, like:
```ini
[fetch]
…
raw_dir = c:\users\gryf\backups\slack\raw
…
```
instead of `raw-dir`. Although when used with commandline switch it has to be `--raw-dir`.

As for the second one (generating with `url_file_to_attachment` set to `true`), you're right. I'm investigating that right now.

**Comment by gryf** [comment]
I've pushed two new commits with sorting things out. could you check it please, @aberja?

**Comment by aberja** [comment]
Thanks for working on this.

I was able to get raw_dir working, but for some reason when I used quotes in the config file the script would not work.  I had to remove the spaces from directory names to allow me to remove the quotes.  i.e. 

[fetch] ... did not work ...
raw_dir = "D:/Documents/_Move/_Move to U/backup/raw_dir"

[fetch] ... does work ...
raw_dir = D:/Documents/_Move/_Move_to_U/backup/raw_dir

The error message was:

2018-09-28 20:42:55,671 INFO: Found configuration file: ./slack-backup.ini
Traceback (most recent call last):
  File "C:/Program Files (x86)/Microsoft Visual Studio/Shared/Anaconda3_64/Scripts/slack-backup", line 10, in <module>
    command.main()
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\slack_backup\command.py", line 143, in main
    args.func(args)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\slack_backup\command.py", line 61, in fetch_data
    slack = client.Client(args)
  File "C:\Program Files (x86)\Microsoft Visual Studio\Shared\Anaconda3_64\lib\site-packages\slack_backup\client.py", line 53, in __init__
    os.mkdir(args.raw_dir)
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: '"D:\\Documents\\_Move\\_Move to U\\backup\\raw_dir"'

I also tried with forward slashes and the error message's last line was:
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: '"D:/Documents/_Move/_Move to U/backup/raw_dir"'


A Couple of questions about raw_dir.
1. Would it be possible to only create files that have content inside the files?  I have a lot of files inside the raw_dir that only include [].

2. You mention that the directory is "where pure response JSONs from Slack API will be stored". Can you elaborate on this?  The files that includes information are hundreds of lines long and  appear to all run together.  It's not clear to me what I'm seeing and what I should be looking for.


I've set url_file_to_attachment = true in the config file, but I cannot locate any file which is prefixed manual_download_ . Where should this file be located and if there is no file does that mean that the script was able to download all attachments and there were none for me to manually download?  If this was the case, would it be possible for the script to note this after running?

**Comment by gryf** [comment]
Not sure how it's working on windows, but if you put quotas on config, you'll get it included in the string, that's how configparser works:
```shell-session
$ cat foo.ini 
[section]
a = /home/gryf/t/one two three/some file.txt
b = "/home/gryf/t/one two three/some file.txt"
c = /home/gryf/t/one\ two\ three/some\ file.txt

$ python3
Python 3.6.5 (default, Jun 20 2018, 17:48:28) 
[GCC 6.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import configparser
>>> import os
>>> parser = configparser.ConfigParser()
>>> parser.read('foo.ini')
['foo.ini']
>>> parser.sections()
['section']
>>> for i in 'a b c'.split():
...     print("option %s: `%s'," % (i, parser.get('section', i)), "exists?:", os.path.exists(parser.get('section', i)))
... 
option a: `/home/gryf/t/one two three/some file.txt', exists?: True
option b: `"/home/gryf/t/one two three/some file.txt"', exists?: False
option c: `/home/gryf/t/one\ two\ three/some\ file.txt', exists?: False
```
So, as you can see, option value is taken literally with everything you throw at it with quotas, backslashes and so on. So if you happen to have a path with quotas in it (i.e. `D:\foo "bar" baz`) it should be placed as is to the ini option, like `some_option = D:\foo "bar" baz`.

As for `raw_dir`.
1. I'm not sure what you mean by "files inside files" `¯\_(ツ)_/¯`. Files containing only `[]` possibly means, that server returned empty list.
2. Slack backup utilizes [slack-client](https://github.com/slackapi/python-slackclient), which underneath use [requests](http://docs.python-requests.org/en/master/) library to communicate with Slack API. Response from the Slack server is simply json structures, which hold of all the data needed to describe requested objects (channel info, messages, user lists and so on). If `raw_dir` is set to some directory, all those Slack server responses will be stored there. Those raw json files might help during debugging, or can be used with another tool.

**Comment by aberja** [comment]
Thanks for the detailed explanation.  It turns out that I don't require quotes in Windows even if there is a space in the file name.

**raw_dir** -  I was referring to the possibility of not creating a file if the server returned an empty list or somehow noting that the file includes an empty list.  This would save one from having to open the files with empty lists.

**url_file_to_attachment** - does not appear to do anything.  At least I cannot find a file with which is prefixed manual_download_ .  

**Files are no longer being downloaded** - I believe the url_file_to_attachment issue may be as a result of a bigger issue...the script is no longer downloading any files.  The last time I had files downloaded was at the end of July.  I've gone into the Slack rooms and confirmed that there are files that should have been downloaded.  To further test this out, I ran the script and had the information dumped into a new folder where images where downloaded to the images folder, but the asset folder was empty.

Looking at the Debug file, it appears that the script is recognizing the files, for example here is a snippet:

 "files": [{"id": "FCDAFXXXX", "created": 1534968551, "timestamp": 1534968551, "name": "StockFundamental.cs", "title": "StockFundamentals.cs", "mimetype": "text/plain", "filetype": "csharp", "pretty_type": "C#", "user": "XXXX", "editable": true, "size": 4132, "mode": "snippet", "is_external": false, "external_type": "", "is_public": true, "public_url_shared": false, "display_as_bot": false, "username": "", "url_private": "https://files.slack.com/files-pri/T5ETEHBDL-XXXX.cs"

**Using Pip to Update script** -  To update the script, I've been copying your modified files to my computer.  However, I'm always concerned that I might miss a file doing it that way.  So I looked into using pip update function and found that the version of slack-backup was not current.  I'm not sure how pip is notified of updates or if it's a lot of work, but it would be great if that was kept updated to allow one to run the pip update slack-backup.

**Comment by gryf** [comment]
> raw_dir - I was referring to the possibility of not creating a file if the 
> server returned an empty list [...]

It's strictly debugging purpose feature, but I think about that. If you think, that it's a must-have behaviour, please fill another issue for it, so we could have an easy tracking.

> url_file_to_attachment - does not appear to do anything. At least I cannot find a file with which is prefixed manual_download_ .

It's work for me though :C Please, make another issue, since it's unrelated to generating a report.

> Files are no longer being downloaded

Yes. That unfortunately truth. I've filled a bug for it #11

> Using Pip to Update script - To update the script, I've been copying your modified files to my computer.

I was waiting to close this issue first, before releasing. It's a common approach, that stable and bugless (ha!) versions are released, while on master/develop branch there could be some rough edges or even major issues.

**Comment by gryf** [comment]
Right. To make the pypi version fixed, #11 have to be fixed first. I'm closing this one, since generating logs should work properly now.
